### PR TITLE
Live Testing - 12/09/24

### PR DESCRIPTION
**Pull request by philmain28**
Test 1 - Initial docker deployment with investigations set up

Commit ID - 29ccdb265d9662207ef01d514f76497bd7d646fe

Date of run - 12/09/24

# Replicating the test
It should be possible to fully replicate this test by cloning the repo and running: 
```
docker compose build 
docker compose up -d 
python scripts/post_files_from_sftp.py 
```

# Report 
Number of uploaded files : 45581
Number of errored files : 2563
Time taken : 1618.42 mins

Errors:
Can\'t instantiate abstract class ProgramMembership with abstract method map_xml_to_orm
Can\'t instantiate abstract class OptOut with abstract method map_xml_to_orm
Upload failed with error: \'ContactDetails\' object is not subscriptable

# Other Issues
- There appear to be a few problems with type 8 issues. I think these should potentially error and not be stored in cupid. Or if they do continue they should have more detail associated with them. ~~The problem here is that a new patient can be created without necessarily being able to insert them. This is something I need to investigate more.~~ It seems like the patientrecord is being created even in the case of a failure which goes against the way I understand sessions (something something atomic transaction).   
-  Some of the "send with range" type things seem to be incomplete. Currently covered are observations and lab orders need to make sure dialysis sessions etc are also covered.  **Done**
-  ~~Check primary key constraints on the lab orders~~  This is a know issue with Devon and Exeter feeds that has since been fixed
-

### CODE REVIEW

**Comment by philmain28** [comment]
Set you all as reviews to records any thoughts you have and make you all aware of what is going on with the repo.

File: tests/test_longitudinal.py [hunk 1]
```diff
@@ -18,63 +18,173 @@
...
-    loaded_ids = [order.id for order in ukrdc_test_session.query(sqla.LabOrder).all()]
+    loaded_ids = [order.id for order in session.query(sqla.LabOrder).all()]
     for order in lab_orders:
         assert order.id in loaded_ids
+    
+    return
+
+
+def generate_persistent_dialysis_sessions(session:Session, start, stop):
+    # Add records to session that are require for testing laborders
```
Comment (George-D-S) [inline tests/test_longitudinal.py:68]: Do the comments on this and generate_persistent_observations need updating to change laborder?

Comment (philmain28) [reply tests/test_longitudinal.py:68]: Ah you spotted the copy pasted code

**Comment by andyatterton** [comment]
I'm wondering if it is possible to get a report from the Jtrace system that mirrors this one. I think you said you were pulling files from staging. Is there an easy way to see how many files failed there as well so we can compare?

Otherwise though good stuff, seems like the testing is going well.

**Review by andyatterton — APPROVED** [review]

**Comment by philmain28** [comment]
That's a good point. I had also thought about some kind of text dump of the records that would allow a diff to be run between the staging and local test ukrdc. Maybe what I am describing is the quarterly extract.

**Review by martabadji — APPROVED** [review]
I can't say I understand much of this but I'm happy there's good progress ;)

**Comment by philmain28** [comment]
> I can't say I understand much of this but I'm happy there's good progress ;)

No problem. I mainly set you as a reviewer for higher level opinions. Hopefully more of which will be required soon.

For example there is currently an investigation raised (type 8) when the file errors on the final step of updating the database. This might be for some reason like: a lab order has been duplicated. For such a file an error could be raised or an investigation. The latter would require some sort of partial insertion to create a patient record.

**Comment by andyatterton** [comment]
This last commit is why we use squash merging 😆

**Comment by George-D-S** [comment]
> That's a good point. I had also thought about some kind of text dump of the records that would allow a diff to be run between the staging and local test ukrdc. Maybe what I am describing is the quarterly extract.

https://renalregistry.atlassian.net/issues/TNG-991 might be an option. Load the files into both then dump the XML. We may need to take account of certain changes if CUPID is operating at 4.2.0 but the Java Repo at 3.5.3.

**Comment by philmain28** [comment]
# Test 2 -  Loading into database with migrated data
The next test is to see what goes wrong with existing ukrdc data migrated across. Hoping this will give me a better opportunity to test the investigations.

**Pull request merged**